File: src/rosilo_robot_kinematics_provider.cpp
#include <rosilo_robot_kinematics/rosilo_robot_kinematics_provider.h>
#include <rosilo_conversions/rosilo_conversions.h>
namespace rosilo
{

void RobotKinematicsProvider::_callback_desired_pose(const geometry_msgs::PoseStamped::ConstPtr &msg)
{
    desired_pose_ = geometry_msgs_pose_to_dq(msg->pose);
    if(! enabled_)
    {
        enabled_ = true;
        ROS_INFO_STREAM(ros::this_node::getName() + "::RobotKinematicsProvider enabled.");
    }
}

RobotKinematicsProvider::RobotKinematicsProvider(ros::NodeHandle &node_handle, const std::string &topic_prefix):
    RobotKinematicsProvider(node_handle, node_handle, topic_prefix)
{
    //Delegated to RobotKinematicsProvider::RobotKinematicsProvider(ros::NodeHandle &nodehandle_publisher, ros::NodeHandle &nodehandle_subscriber, const std::string &topic_prefix)
}

RobotKinematicsProvider::RobotKinematicsProvider(ros::NodeHandle &nodehandle_publisher, ros::NodeHandle &nodehandle_subscriber, const std::string &topic_prefix):
    enabled_(false),
    topic_prefix_(topic_prefix),
    desired_pose_(0)
{
    ROS_INFO_STREAM(ros::this_node::getName() + "::Initializing RobotKinematicsProvider with prefix " + topic_prefix);
    subscriber_desired_pose_ = nodehandle_subscriber.subscribe(topic_prefix + "/set/desired_pose", 1, &RobotKinematicsProvider::_callback_desired_pose, this);
    publisher_pose_ = nodehandle_publisher.advertise<geometry_msgs::PoseStamped>(topic_prefix + "/get/pose", 1);
}

bool RobotKinematicsProvider::is_enabled() const
{
    return enabled_;
}

DQ RobotKinematicsProvider::get_desired_pose() const
{
    if(enabled_)
    {
        return desired_pose_;
    }
    else
    {
        throw std::runtime_error(ros::this_node::getName() + "::Trying to get_desired_pose of an unitialized RobotKinematicsProvider.");
    }
}

void RobotKinematicsProvider::send_pose(const DQ &pose) const
{
    publisher_pose_.publish(dq_to_geometry_msgs_pose_stamped(pose));
}

}

File: src/rosilo_robot_kinematics/rosilo_robot_kinematics_provider.py
"""
# Copyright (c) 2012-2021 Murilo Marques Marinho
#
#    This file is part of rosilo_robot_driver.
#
#    rosilo_robot_driver is free software: you can redistribute it and/or modify
#    it under the terms of the GNU Lesser General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    rosilo_robot_driver is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Lesser General Public License for more details.
#
#    You should have received a copy of the GNU Lesser General Public License
#    along with rosilo_robot_driver.  If not, see <https://www.gnu.org/licenses/>.
#
# ################################################################
#
#   Author: Murilo M. Marinho, email: murilo@nml.t.u-tokyo.ac.jp
#
# ################################################################
"""
import rospy
from geometry_msgs.msg import PoseStamped
from std_msgs.msg import Float64
import rosilo_conversions as rc


class RobotKinematicsProvider:

    def __init__(self, topic_prefix):
        self.desired_interpolator_speed_ = None
        self.desired_pose_ = None
        self.enabled_ = False
        self.topic_prefix = topic_prefix
        self.desired_pose_ = 0

        print(rospy.get_name() + "::Initializing RobotKinematicsProvider with prefix " + topic_prefix)
        self.publisher_pose_ = rospy.Publisher(topic_prefix + "get/pose", PoseStamped, queue_size=1)

        self.subscriber_desired_pose_ = rospy.Subscriber(topic_prefix + "set/desired_pose",
                                                         PoseStamped,
                                                         self._callback_desired_pose)
        self.subscriber_desired_interpolator_speed_ = rospy.Subscriber(topic_prefix + "set/desired_interpolator_speed",
                                                                       Float64,
                                                                       self._callback_desired_interpolator_speed)

    def _callback_desired_pose(self, msg):
        self.desired_pose_ = rc.geometry_msgs_pose_to_dq(msg.pose)
        if not self.enabled_:
            self.enabled_ = True
            print(rospy.get_name() + "::RobotKinematicsProvider enabled.")

    def _callback_desired_interpolator_speed(self, msg):
        self.desired_interpolator_speed_ = rc.std_msgs_float64_to_double(msg)

    def is_enabled(self):
        return self.enabled_

    def get_desired_pose(self):
        if self.enabled_:
            return self.desired_pose_
        else:
            raise Exception(
                rospy.get_name() + "::Trying to get_desired_pose of an unitialized RobotKinematicsProvider.")

    def get_desired_interpolator_speed(self):
        return self.desired_interpolator_speed_

    def send_pose(self, pose):
        self.publisher_pose_.publish(rc.dq_to_geometry_msgs_pose_stamped(pose))
